### PR TITLE
refactor(routing): Api v2 for routing create and activate endpoints

### PR DESCRIPTION
**Pull request by Aprabhat19**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
This PR adds a `Routing_v2` flag and renames all the endpoints, and refactors the create and activate endpoints

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
 
 - Create routing 
 ```
 curl --location 'http://localhost:8080/routing_algorithm' \
--header 'Content-Type: application/json' \
--header 'api-key: dev_1ZYFaypqi4RKpkjTFbHHkdgAqvUsG6TYJXoH53D59BQ1uFMrBIcxrF4f7FiJkOWw' \
--data '{
    "name": "newalgo",
    "description": "It is my ADVANCED config",
    "algorithm": {
        "type": "advanced",
        "data": {
            "defaultSelection": {
                "type": "priority",
                "data": [
                   {"connector": "adyen",
                    "merchant_connector_id":"mca_XcrOMvDBAiq6ln7ESbu0"}
                ]
            },
            "rules": [
                {
                    "name": "adyen",
                    "connectorSelection": {
                        "type": "priority",
                        "data": [
                       { "connector": "adyen",
                    "merchant_connector_id":"mca_XcrOMvDBAiq6ln7ESbu0"},
                    {
                        "connector": "cybersource",
                    "merchant_connector_id":"mca_8uCqbT9Bt9bmKUVuR3IA"
                    }

                        ]
                    },
                    "statements": [
                        {
                            "condition": [
                                {
                                    "lhs": "payment_amount",
                                    "comparison": "greater_than",
                                    "value": {
                                        "type": "number",
                                        "value": 100077
                                    },
                                    "metadata": {}
                                }
                            ],
                            "nested": null
                        }
                    ]
                }
            ],
            "metadata": {}
        }
    },
    "profile_id":"pro_r6qEFW0XLe7UnQoaaQ67"
}'
```
- Activate Routing Algorihtm
```
curl --location --request PATCH 'http://localhost:8080/v2/profiles/pro_r6qEFW0XLe7UnQoaaQ67/activate_routing_algorithm' \
--header 'Content-Type: application/json' \
--header 'api-key: dev_1ZYFaypqi4RKpkjTFbHHkdgAqvUsG6TYJXoH53D59BQ1uFMrBIcxrF4f7FiJkOWw' \
--data '"routing_merchant_1721842459_WWHyZVY3oW"
'
```
- The activated algorithm is then store in Business Profile Against that profile_id
Screenshot 2024-07-29 at 2 33 56 PM

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — APPROVED** [review]

File: crates/router/src/core/routing.rs [hunk 1]
```diff
@@ -75,8 +154,96 @@ pub async fn create_routing_config(
     transaction_type: &enums::TransactionType,
 ) -> RouterResponse<routing_types::RoutingDictionaryRecord> {
     metrics::ROUTING_CREATE_REQUEST_RECEIVED.add(&metrics::CONTEXT, 1, &[]);
-    let db = state.store.as_ref();
+    let db = &*state.store;
+    let name = request
```
Comment (Narayanbhat166) [inline crates/router/src/core/routing.rs:158]: if this is a required value, then have it as required in the api model right?

File: crates/router/src/routes/app.rs [hunk 2]
```diff
@@ -567,15 +570,35 @@ impl Forex {
 #[cfg(feature = "olap")]
 pub struct Routing;
 
-#[cfg(feature = "olap")]
+#[cfg(all(feature = "olap", feature = "v2", feature = "routing_v2"))]
+impl Routing {
+    pub fn server(state: AppState) -> Scope {
+        web::scope("/routing_algorithm")
```
Comment (Narayanbhat166) [inline crates/router/src/routes/app.rs:576]: prefix should be v2

File: crates/router/src/routes/app.rs [hunk 3]
```diff
@@ -1361,8 +1367,64 @@ impl PayoutLink {
 }
 
 pub struct BusinessProfile;
-
-#[cfg(feature = "olap")]
+#[cfg(all(feature = "olap", feature = "v2", feature = "routing_v2"))]
+impl BusinessProfile {
+    pub fn server(state: AppState) -> Scope {
+        web::scope("profiles")
```
Comment (Narayanbhat166) [inline crates/router/src/routes/app.rs:1373]: even here, the prefix should be v2

**Review by Narayanbhat166 — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

File: crates/router/src/core/routing.rs [hunk 4]
```diff
@@ -148,6 +313,56 @@ pub async fn create_routing_config(
...
+
+    let mut routing_ref = routing_types::RoutingAlgorithmRef::parse_routing_algorithm(
+        business_profile.routing_algorithm.clone().map(Secret::new),
+    )
+    .change_context(errors::ApiErrorResponse::InternalServerError)
+    .attach_printable("unable to deserialize routing algorithm ref from merchant account")?
+    .unwrap_or_default();
+
+    routing_algorithm.update_routing_ref_with_algorithm_id(transaction_type, &mut routing_ref)?;
+    // TODO move to business profile
```
Comment (prajjwalkumar17) [inline crates/router/src/core/routing.rs:351]: Is this needed?

File: crates/router/src/core/routing/helpers.rs [hunk 5]
```diff
@@ -238,7 +170,7 @@ pub async fn update_merchant_active_algorithm_ref(
 
     Ok(())
 }
-
+// TODO: Move it to business_profile
```
Comment (prajjwalkumar17) [inline crates/router/src/core/routing/helpers.rs:173]: is this needed?

**Review by prajjwalkumar17 — CHANGES_REQUESTED** [review]
Else everything LGTM!

**Review by prajjwalkumar17 — APPROVED** [review]

File: crates/router/src/core/routing.rs (hunk 4 above)
Comment (Aprabhat19) [reply crates/router/src/core/routing.rs:351]: Yes, Its supposed to be handled later on , due to inter dependency on BusinessProfile

File: crates/router/src/core/routing/helpers.rs (hunk 5 above)
Comment (Aprabhat19) [reply crates/router/src/core/routing/helpers.rs:173]: Same its supposed to be handled later on

**Review by lsampras — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Pull request merged**